### PR TITLE
Introduce gosec for security checks

### PR DESCRIPTION
**Pull request by JAORMX**
gosec does static code analysis and checks for common security issues in
golang codebases.

This PR introduces the tool, and sets it up as a test target in the Makefile.

### CODE REVIEW

**Comment by ironcladlou** [comment]
As to whether we want to accept this generally: are we promoting this for all OpenShift component builds? cc @sttts @smarterclayton 

If we do want this tool to gate our builds, I will ask that you please refactor to incorporate the tool into our vending system ([see here](https://github.com/openshift/cluster-ingress-operator/blob/master/tools/tools.go) and [here](https://github.com/openshift/cluster-ingress-operator/blob/master/hack/update-generated-bindata.sh#L8)). We require this technique primarily for build reproducibility.

**Comment by JAORMX** [comment]
@ironcladlou I'm unsure on what you mean by including this in the component builds.

This tool won't change anything in the code, nor the component. All it does is do static analysis on the code and report security related warnings.

**Comment by ironcladlou** [comment]
>@ironcladlou I'm unsure on what you mean by including this in the component builds.
>This tool won't change anything in the code, nor the component. All it does is do static analysis on the code and report security related warnings.

What I mean is, what's the background and expectation across all our core operators? Should a failure of this check block merges? If not, what's the point? And if so, are we going to apply this merge gate to _every core operator_? Will you be opening PRs for all of them, or asking teams to open them? If not, why just this operator? Was there some discussion I missed about this particular choice of analysis?

I'm not necessarily opposed, I just need some context.

The refactoring described in https://github.com/openshift/cluster-ingress-operator/pull/276#issuecomment-518281130 will still be necessary if we decide to introduce the check.

**Comment by JAORMX** [comment]
@ironcladlou I've been proposing this check all over the repos in openshift. Still several repos to go. Some have accepted it, some are still pending reviewing this. There hasn't been a particular dicussion where this was approved. I'm looking at certifications for OpenShift (FedRAMP in particular); and having security checks will address one of the security controls mandated by that.

**Comment by JAORMX** [comment]
Any more feedback on this PR?

**Comment by ironcladlou** [comment]
@JAORMX https://github.com/openshift/cluster-ingress-operator/pull/276#issuecomment-518281130

It needs refactored to use vendoring.

**Comment by JAORMX** [comment]
> @JAORMX [#276 (comment)](https://github.com/openshift/cluster-ingress-operator/pull/276#issuecomment-518281130)
> 
> It needs refactored to use vendoring.

Can you elaborate on this? I'm not sure what you mean

File: Makefile [hunk 1]
```diff
@@ -62,9 +62,13 @@ clean:
 	rm -f $(BIN)
 
 .PHONY: verify
-verify: verify-bindata verify-crd
+verify: verify-bindata verify-crd sec
 	hack/verify-gofmt.sh
 
+.PHONY: sec
+sec:
+	hack/gosec.sh
```
[the remark this answers is not in the document]
Comment (JAORMX) [reply Makefile]: Done. thanks for the feedback.

File: Makefile [hunk 2]
```diff
@@ -62,9 +62,13 @@ clean:
 	rm -f $(BIN)
 
 .PHONY: verify
-verify: verify-bindata verify-crd
+verify: verify-bindata verify-crd sec
 	hack/verify-gofmt.sh
 
+.PHONY: sec
```
[the remark this answers is not in the document]
Comment (JAORMX) [reply Makefile]: Done

File: hack/verify-gosec.sh [hunk 3]
```diff
@@ -0,0 +1,4 @@
+#!/bin/bash
+set -euo pipefail
+
+GO111MODULE=on GOFLAGS=-mod=vendor go run github.com/securego/gosec/cmd/gosec -severity medium -confidence medium -exclude G304 -quiet ./...
```
[the remark this answers is not in the document]
Comment (JAORMX) [reply hack/verify-gosec.sh]: So... I've been proposing this to a bunch of openshift repos already. And time and time again folks ask about the warnings that come from G304; with every time the teams realizing that it's not a very applicable check to this type of code-base. So, I decided to exclude it by default.

[the remark this answers is not in the document]
Comment (JAORMX) [reply hack/verify-gosec.sh]: ok; done

**Pull request merged**